File: runtimes/jvm/src/com/flint/Conc.java
package com.flint;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;

/// The concurrency builtins: green threads and LOCAL channels
/// (`doc/decisions/0005`).
///
/// Local only, deliberately. A channel joins two green threads inside one
/// sandbox and passes values BY REFERENCE — same heap, same collector, nothing
/// to protect either end from. A global port crosses a heap and is the host's
/// (`doc/decisions/0027`); it is a different object with a different cost and
/// is not built here yet.
public final class Conc {
    private Conc() {}

    /// A local channel end. Two of them, each holding the other.
    public static final class Chan {
        public final long id;
        public final int cap;
        public final String label;
        public final ArrayDeque<Object> inbox = new ArrayDeque<>();
        public Chan peer;
        public boolean closed;
        Chan(long id, int cap, String label) { this.id = id; this.cap = cap; this.label = label; }
        @Override public String toString() {
            return "#<port " + id + (label == null ? "" : " " + label) + ">";
        }
    }

    private static long nextChan = 1;

    static void install() {
        Builtins.def("flint/spawn", (vm, a) -> {
            Green.Thread t = new Green.Thread(vm.sched.nextId++, a[0], null);
            vm.sched.threads.add(t);
            return t;
        });
        Builtins.def("flint/self", (vm, a) -> vm.sched.cur());
        Builtins.def("flint/thread?", (vm, a) -> a[0] instanceof Green.Thread);
        Builtins.def("flint/thread-id", (vm, a) -> ((Green.Thread) a[0]).id);
        Builtins.def("flint/thread-state", (vm, a) -> Kw.of(null, stateName(((Green.Thread) a[0]).status)));
        Builtins.def("flint/thread-result", (vm, a) -> ((Green.Thread) a[0]).result);

        /// `yield` is a COURTESY yield: the call has already happened, so it
        /// must not be re-executed on resume. It parks on nothing and is made
        /// runnable again immediately, which lets every other runnable thread
        /// have a turn before this one continues.
        Builtins.def("flint/yield", (vm, a) -> {
            Green.Thread t = vm.sched.cur();
            if (t == null || vm.sched.threads.size() < 2) return null;
            throw vm.yieldNow();
        });

        Builtins.def("flint/thread-join", (vm, a) -> {
            Green.Thread t = (Green.Thread) a[0];
            if (t.status == Green.DONE) return t.result;
            if (t.status == Green.FAILED) throw new Vm.Thrown(t.result);
            Green.Thread me = vm.sched.cur();
            if (me == t) throw new Vm.Thrown("a thread cannot join itself");
            throw vm.park(t);
        });

        Builtins.def("flint/channel", (vm, a) -> {
            int cap = a.length > 0 && a[0] instanceof Long n ? n.intValue() : 16;
            String label = a.length > 1 && a[1] instanceof String s ? s : null;
            Chan x = new Chan(nextChan++, cap, label);
            Chan y = new Chan(nextChan++, cap, label);
            x.peer = y; y.peer = x;
            List<Object> pair = new ArrayList<>();
            pair.add(x); pair.add(y);
            return pair;
        });
        Builtins.def("flint/port?", (vm, a) -> a[0] instanceof Chan);
        Builtins.def("flint/port-id", (vm, a) -> ((Chan) a[0]).id);
        Builtins.def("flint/port-label", (vm, a) -> ((Chan) a[0]).label);
        Builtins.def("flint/port-host?", (vm, a) -> false);
        Builtins.def("flint/port-state", (vm, a) -> {
            Chan c = (Chan) a[0];
            return Kw.of(null, c.closed ? "closed" : (c.peer.closed && c.inbox.isEmpty() ? "closed" : "open"));
        });
        Builtins.def("flint/port-close", (vm, a) -> {
            Chan c = (Chan) a[0];
            c.closed = true;
            vm.wakeOn(c);
            vm.wakeOn(c.peer);
            return null;
        });

        /// A send goes into the PEER's inbox, which is what makes a pair
        /// two-way. Parks when that inbox is full, and the receiver draining it
        /// is what wakes the sender.
        Builtins.def("flint/port-send", (vm, a) -> {
            Chan c = (Chan) a[0];
            Chan to = c.peer;
            if (to.closed) throw new Vm.Thrown(
                "the other end of this port is gone, so nothing can ever receive this");
            if (to.inbox.size() >= to.cap) throw vm.park(to);
            to.inbox.add(a[1]);
            vm.wakeOn(to);
            return null;
        });

        /// A receive takes from this end's OWN inbox. Parks when it is empty
        /// and the peer is still open; reads as end-of-stream once the peer has
        /// closed and everything buffered is drained.
        Builtins.def("flint/port-receive", (vm, a) -> {
            Chan c = (Chan) a[0];
            if (!c.inbox.isEmpty()) {
                Object v = c.inbox.poll();
                vm.wakeOn(c);
                return v;
            }
            if (c.peer.closed || c.closed) return null;
            throw vm.park(c);
        });

        // Dynamic bindings are per green thread, and a spawn inherits a
        // snapshot. Not built here yet: the JVM port keeps them per REAL
        // thread (`Builtins.dyn*`), which is right for its own threading and
        // wrong for flint's. Named rather than silently left.
    }

    private static String stateName(int st) {
        switch (st) {
            case Green.NEW: return "new";
            case Green.RUNNABLE: return "runnable";
            case Green.PARKED: return "parked";
            case Green.DONE: return "done";
            default: return "failed";
        }
    }
}

File: runtimes/clr/src/Conc.cs
using System.Collections.Generic;

namespace Flint;

/// The concurrency builtins: green threads and LOCAL channels
/// (`doc/decisions/0005`). The CLR half of what `Conc.java` does, kept
/// deliberately line-for-line with it so the two agree by construction rather
/// than by testing.
///
/// Local only. A channel joins two green threads inside one sandbox and passes
/// values BY REFERENCE -- same heap, nothing to protect either end from. A
/// global port crosses a heap, is the host's, and is a different object
/// (`doc/decisions/0027`); it is not built here yet.
public static class Conc {
    public sealed class Chan {
        public readonly long Id;
        public readonly int Cap;
        public readonly string Label;
        public readonly Queue<object> Inbox = new();
        public Chan Peer;
        public bool Closed;
        public Chan(long id, int cap, string label) { Id = id; Cap = cap; Label = label; }
        public override string ToString() =>
            "#<port " + Id + (Label == null ? "" : " " + Label) + ">";
    }

    private static long _nextChan = 1;

    public static void Install(System.Action<string, Builtins.Fn> def) {
        def("flint/spawn", (vm, a) => {
            var t = new Green.Thread(vm.Sched.NextId++, a[0]);
            vm.Sched.Threads.Add(t);
            return t;
        });
        def("flint/self", (vm, a) => vm.Sched.Cur);
        def("flint/thread?", (vm, a) => a[0] is Green.Thread);
        def("flint/thread-id", (vm, a) => ((Green.Thread) a[0]).Id);
        def("flint/thread-state", (vm, a) => Kw.Of(null, StateName(((Green.Thread) a[0]).Status)));
        def("flint/thread-result", (vm, a) => ((Green.Thread) a[0]).Result);

        def("flint/yield", (vm, a) => {
            var t = vm.Sched.Cur;
            if (t == null || vm.Sched.Threads.Count < 2) return null;
            throw vm.YieldNow();
        });

        def("flint/thread-join", (vm, a) => {
            var t = (Green.Thread) a[0];
            if (t.Status == Green.Done) return t.Result;
            if (t.Status == Green.Failed) throw new FlintThrow(t.Result);
            if (ReferenceEquals(vm.Sched.Cur, t)) throw new FlintThrow("a thread cannot join itself");
            throw vm.ParkOn(t);
        });

        def("flint/channel", (vm, a) => {
            int cap = a.Length > 0 && a[0] is long n ? (int) n : 16;
            string label = a.Length > 1 ? a[1] as string : null;
            var x = new Chan(_nextChan++, cap, label);
            var y = new Chan(_nextChan++, cap, label);
            x.Peer = y; y.Peer = x;
            return new Vec(new List<object> { x, y });
        });
        def("flint/port?", (vm, a) => a[0] is Chan);
        def("flint/port-id", (vm, a) => ((Chan) a[0]).Id);
        def("flint/port-label", (vm, a) => ((Chan) a[0]).Label);
        def("flint/port-host?", (vm, a) => false);
        def("flint/port-state", (vm, a) => {
            var c = (Chan) a[0];
            return Kw.Of(null, c.Closed || (c.Peer.Closed && c.Inbox.Count == 0) ? "closed" : "open");
        });
        def("flint/port-close", (vm, a) => {
            var c = (Chan) a[0];
            c.Closed = true;
            vm.WakeOn(c);
            vm.WakeOn(c.Peer);
            return null;
        });

        /// A send goes into the PEER's inbox, which is what makes a pair
        /// two-way. Parks on a full inbox; the receiver draining it wakes us.
        def("flint/port-send", (vm, a) => {
            var c = (Chan) a[0];
            var to = c.Peer;
            if (to.Closed)
                throw new FlintThrow("the other end of this port is gone, so nothing can ever receive this");
            if (to.Inbox.Count >= to.Cap) throw vm.ParkOn(to);
            to.Inbox.Enqueue(a[1]);
            vm.WakeOn(to);
            return null;
        });

        /// A receive takes from this end's OWN inbox. Parks when empty and the
        /// peer is open; reads as end-of-stream once the peer has closed and
        /// everything buffered is drained.
        def("flint/port-receive", (vm, a) => {
            var c = (Chan) a[0];
            if (c.Inbox.Count > 0) {
                var v = c.Inbox.Dequeue();
                vm.WakeOn(c);
                return v;
            }
            if (c.Peer.Closed || c.Closed) return null;
            throw vm.ParkOn(c);
        });
    }

    private static string StateName(int st) => st switch {
        Green.New => "new",
        Green.Runnable => "runnable",
        Green.Parked => "parked",
        Green.Done => "done",
        _ => "failed",
    };
}
